### PR TITLE
refactor(expr): introduce ExprError

### LINKED ISSUE #3082: Introduce ExprError Type

### PR DESCRIPTION
**Pull request by TennyZhuang**
Signed-off-by: TennyZhuang <zty0826@gmail.com>

## What's changed and what's your intention?

***PLEASE DO NOT LEAVE THIS EMPTY !!!***

As title

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)

close #3082

### CODE REVIEW

File: src/common/src/error.rs [hunk 1]
```diff
@@ -96,6 +96,12 @@ pub enum ErrorCode {
         #[source]
         BoxedError,
     ),
+    #[error("Expr error: {0:?}")]
+    ExprError(
```
Comment (TennyZhuang) [inline src/common/src/error.rs:100]: I'm not sure should we introduce `ExprError(ExprError)` here instead of convert to other RwError.

**Review by BowenXiao1999 — APPROVED** [review]
LGTM!

Looking forward to it!

File: src/common/src/error.rs (hunk 1 above)
Comment (fuyufjh) [reply src/common/src/error.rs:100]: > pub type BoxedError = Box<dyn Error + Send + Sync>;

So there the error is boxed into a trait object. 🤔

`ExprError(ExprError)` looks better to me.

Comment (BugenZhao) [reply src/common/src/error.rs:100]: I think the expression error will always be capsulated by some other error like `StreamError` when exposed to the user, so there might be no need to make it a variant of `RwError` eventually.
We may keep it for now due to compatibility.

File: src/expr/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    Parse(&'static str),
+
+    #[error("Invalid parameter {name}: {reason}")]
+    InvalidParam { name: &'static str, reason: String },
+
+    #[error("Array error: {0}")]
+    Array(
+        #[backtrace]
+        #[source]
+        RwError,
```
Comment (fuyufjh) [inline src/expr/src/error.rs:45]: Will we refactor this later?

File: src/expr/src/vector_op/agg/general_agg.rs [hunk 3]
```diff
@@ -256,7 +256,7 @@ mod tests {
                 Arc::new(input.into()),
                 &agg_type,
                 return_type,
-                ArrayBuilderImpl::Int32(I32ArrayBuilder::new(0)?),
+                ArrayBuilderImpl::Int32(I32ArrayBuilder::new(0).unwrap()),
```
Comment (fuyufjh) [inline src/expr/src/vector_op/agg/general_agg.rs:259]: Why `ArrayBuilder::new` returns `Result`...

**Review by fuyufjh — COMMENTED** [review]
LGTM

File: src/expr/src/error.rs (hunk 2 above)
Comment (BugenZhao) [inline src/expr/src/error.rs:45]: May review this in the future. 🤣

**Review by BugenZhao — APPROVED** [review]
Cool!

File: src/common/src/error.rs (hunk 1 above)
Comment (BugenZhao) [reply src/common/src/error.rs:100]: > > pub type BoxedError = Box<dyn Error + Send + Sync>;
> 
> So there the error is boxed into a trait object. 🤔
> 
> `ExprError(ExprError)` looks better to me.

It seems we can't refer `ExprError` here due to crate dependencies?

Comment (TennyZhuang) [reply src/common/src/error.rs:100]: > > > pub type BoxedError = Box<dyn Error + Send + Sync>;
> > 
> > 
> > So there the error is boxed into a trait object. 🤔
> > `ExprError(ExprError)` looks better to me.
> 
> It seems we can't refer `ExprError` here due to crate dependencies?

Yes, same as `StorageError`

File: src/expr/src/error.rs (hunk 2 above)
Comment (TennyZhuang) [reply src/expr/src/error.rs:45]: Yes

File: src/expr/src/vector_op/agg/general_agg.rs (hunk 3 above)
Comment (TennyZhuang) [reply src/expr/src/vector_op/agg/general_agg.rs:259]: Allocation may failed 😃

File: src/expr/src/error.rs (hunk 2 above)
Comment (TennyZhuang) [reply src/expr/src/error.rs:45]: We need ArrayError :(

File: src/expr/src/vector_op/agg/general_agg.rs (hunk 3 above)
Comment (fuyufjh) [reply src/expr/src/vector_op/agg/general_agg.rs:259]: Anyway, here it's `unwrap`ed.

Comment (TennyZhuang) [reply src/expr/src/vector_op/agg/general_agg.rs:259]: Here is tests

**Pull request merged**